### PR TITLE
Check for duplicate label names in remote read

### PR DESCRIPTION
**Pull request by vmartsynovskyy**
In #5731 a bug was discovered where there would sometimes be a timeseries from local storage and a timeseries from a remoteRead that were duplicates of eachother. This happened in cases where a timeseries coming from a remoteRead had duplicate label names. This caused `mergeSeriesSet` to not merge the remoteRead timeseries with the local storage timeseries.

This PR fixes this issue by adding a function to check for and remove duplicate label names coming from a remoteRead. There are also tests that ensure this is really happening.

CC @brian-brazil

### CODE REVIEW

File: storage/remote/codec.go [hunk 1]
```diff
@@ -352,6 +353,21 @@ func LabelProtosToMetric(labelPairs []*prompb.Label) model.Metric {
 	return metric
 }
 
+// uniqueLabelNameProtos creates a []prompb.Label with duplicate label names removed
```
Comment (brian-brazil) [inline storage/remote/codec.go]: We should error in this case, not hide the issue.

Comment (vmartsynovskyy) [reply storage/remote/codec.go]: Good point. I'll add a check for duplicate labels in `validateLabelsAndMetricName` and update the tests I wrote accordingly.

File: storage/remote/codec_test.go [hunk 2]
```diff
@@ -145,3 +147,59 @@ func TestConcreteSeriesClonesLabels(t *testing.T) {
 	gotLabels = cs.Labels()
 	require.Equal(t, lbls, gotLabels)
 }
+
+func TestDuplicateLabelsSimple(t *testing.T) {
```
Comment (cstyan) [inline storage/remote/codec_test.go]: you could turn this and `TestDuplicateLabelsDoNothing` into a single test by using a subtest and turning the input/expected output into a struct

Comment (vmartsynovskyy) [reply storage/remote/codec_test.go]: I removed these tests and replaced them with more test cases in `TestValidateLabelsAndMetricName`. I'll see if I can refactor that test to use subtests though.

Comment (vmartsynovskyy) [reply storage/remote/codec_test.go]: I refactored `TestValidateLabelsAndMetricName` to use subtests, let me know if it looks right to you.

File: storage/remote/codec.go [hunk 3]
```diff
@@ -277,8 +277,10 @@ func (c *concreteSeriesIterator) Err() error {
 	return nil
 }
 
-// validateLabelsAndMetricName validates the label names/values and metric names returned from remote read.
+// validateLabelsAndMetricName validates the label names/values and metric names returned from remote read,
+// also making sure that there are no labels with duplicate names
 func validateLabelsAndMetricName(ls labels.Labels) error {
+	existingLabelNames := make(map[string]bool, len(ls))
```
Comment (brian-brazil) [inline storage/remote/codec.go]: Labels will be in order, so you only need to compare with the previous one.

Comment (vmartsynovskyy) [reply storage/remote/codec.go]: Nice catch. I didn't notice that `labelProtosToLabels` sorts the labels.

I updated the validation function and added a corresponding test case (as well as updating the tests to actually check that expectedErr matches err).

File: storage/remote/codec.go [hunk 4]
```diff
@@ -289,6 +291,10 @@ func validateLabelsAndMetricName(ls labels.Labels) error {
 		if !model.LabelValue(l.Value).IsValid() {
 			return errors.Errorf("invalid label value: %v", l.Value)
 		}
+		if l.Name == prevLabelName {
```
Comment (brian-brazil) [inline storage/remote/codec.go]: You could avoid having prevLabelName by getting the index from the for, checking that it's >0 and then looking at `ls[i-1]`

Comment (vmartsynovskyy) [reply storage/remote/codec.go]: Sure, I guess its cleaner that way.

File: storage/remote/codec_test.go [hunk 5]
```diff
@@ -44,57 +48,96 @@ func TestValidateLabelsAndMetricName(t *testing.T) {
...
 		},
 		{
 			input: labels.FromStrings(
 				"__name__", "name",
 				"labelName", string([]byte{0xff}),
 			),
-			expectedErr: "Invalid label value: " + string([]byte{0xff}),
+			expectedErr: "invalid label value: " + string([]byte{0xff}),
 			shouldPass:  false,
+			description: "label value is 0xff",
```
Comment (brian-brazil) [inline storage/remote/codec_test.go]: Invalid UTF=8 value

Comment (vmartsynovskyy) [reply storage/remote/codec_test.go]: Updated

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**